### PR TITLE
feat(amplify-appsync-simulator): implement GraphQL Info Object

### PR DESCRIPTION
**Pull request by bboure**
*Issue #, if available:*
None

*Description of changes:*
Implements the GraphQL Info Object as described [here](https://aws.amazon.com/blogs/mobile/appsync-and-the-graphql-info-object/).

Note: This implementation is a reverse engineering using the info I could find in the aforementioned post and by testing the real thing in AppSync.
Any help, including for testing it, would be appreciated.

Todo:
- [x] Add unit tests

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yuth** [comment]
Thank you for the pull request. I see that your change does not address when the section set has a fragment spread which is supported by the [AppSync service](https://docs.aws.amazon.com/appsync/latest/devguide/resolver-context-reference.html#aws-appsync-resolver-context-reference-info).

File: packages/amplify-appsync-simulator/src/velocity/util/info.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { GraphQLResolveInfo, SelectionNode } from 'graphql';
+import { print } from 'graphql/language/printer';
+
+const getSelectionSet = (nodes: readonly SelectionNode[], prefix: string = null) => {
```
Comment (yuth) [inline packages/amplify-appsync-simulator/src/velocity/util/info.ts]: Could you add some tests to test cases where prefix is passed. Also could you add support for aliases

**Comment by bboure** [comment]
@yuth 
> Thank you for the pull request. I see that your change does not address when the section set has a fragment spread which is supported by the AppSync service.

I forgot that use case and am working on a fix. 
Quick question: 

I was wondering is there is an issue in the doc:

````graphql
    ... on Post {
      inlineFrag: comments: {
        id
      }
    }
    ... postFrag
````

according to the doc, becomes:

````js
"inlineFragComments",
"inlineFragComments/id",
````

But when I try this in appsync I get

````js
"inlineFrag",
"inlineFrag/id",
````

The alias is used, and not a concatenation of the alias and the name as in the doc.

**Comment by yuth** [comment]
# [Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=h1) Report
> Merging [#3450](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=desc) into [master](https://codecov.io/gh/aws-amplify/amplify-cli/commit/ad3a1375ff1e81a6ef5567c2518434114c852df0&el=desc) will **increase** coverage by `0.44%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master    #3450      +/-   ##
==========================================
+ Coverage   59.68%   60.13%   +0.44%     
==========================================
  Files         283      280       -3     
  Lines       13143    13079      -64     
  Branches     2641     2488     -153     
==========================================
+ Hits         7845     7865      +20     
+ Misses       4876     4809      -67     
+ Partials      422      405      -17     
```


| [Impacted Files](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [packages/graphql-mapping-template/src/print.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvZ3JhcGhxbC1tYXBwaW5nLXRlbXBsYXRlL3NyYy9wcmludC50cw==) | `35.71% <0.00%> (ø)` | :arrow_up: |
| [...ages/amplify-appsync-simulator/src/server/index.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS1hcHBzeW5jLXNpbXVsYXRvci9zcmMvc2VydmVyL2luZGV4LnRz) | `64.28% <0.00%> (ø)` | :arrow_up: |
| [...ges/amplify-util-mock/src/CFNParser/stack/index.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL0NGTlBhcnNlci9zdGFjay9pbmRleC50cw==) | `88.09% <0.00%> (ø)` | :arrow_up: |
| [...ges/graphql-transformer-core/src/util/fileUtils.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvZ3JhcGhxbC10cmFuc2Zvcm1lci1jb3JlL3NyYy91dGlsL2ZpbGVVdGlscy50cw==) | `11.76% <0.00%> (ø)` | :arrow_up: |
| [...es/graphql-transformer-core/src/stripDirectives.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvZ3JhcGhxbC10cmFuc2Zvcm1lci1jb3JlL3NyYy9zdHJpcERpcmVjdGl2ZXMudHM=) | `35.29% <0.00%> (ø)` | :arrow_up: |
| [...amplify-appsync-simulator/src/server/operations.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS1hcHBzeW5jLXNpbXVsYXRvci9zcmMvc2VydmVyL29wZXJhdGlvbnMudHM=) | `28.45% <0.00%> (ø)` | :arrow_up: |
| [...graphql-transformer-core/src/TransformerContext.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvZ3JhcGhxbC10cmFuc2Zvcm1lci1jb3JlL3NyYy9UcmFuc2Zvcm1lckNvbnRleHQudHM=) | `26.20% <0.00%> (ø)` | :arrow_up: |
| [...mplify-appsync-simulator/src/resolvers/function.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS1hcHBzeW5jLXNpbXVsYXRvci9zcmMvcmVzb2x2ZXJzL2Z1bmN0aW9uLnRz) | `9.67% <0.00%> (ø)` | :arrow_up: |
| [...plify-appsync-simulator/src/server/subscription.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS1hcHBzeW5jLXNpbXVsYXRvci9zcmMvc2VydmVyL3N1YnNjcmlwdGlvbi50cw==) | `26.95% <0.00%> (ø)` | :arrow_up: |
| [...lify-graphql-types-generator/src/compiler/index.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS1ncmFwaHFsLXR5cGVzLWdlbmVyYXRvci9zcmMvY29tcGlsZXIvaW5kZXgudHM=) | `90.32% <0.00%> (ø)` | :arrow_up: |
| ... and [16 more](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=footer). Last update [ad3a137...a5b65a2](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3450?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: packages/amplify-appsync-simulator/src/velocity/util/info.ts (hunk 1 above)
Comment (bboure) [reply packages/amplify-appsync-simulator/src/velocity/util/info.ts]: @yuth Support for alises added. 
test for prefix is included in test.

**Comment by yuth** [comment]
> I was wondering is there is an issue in the doc:

I think so. I will get in touch with the service team to update the docs. Thank you

**Pull request merged**